### PR TITLE
Remote Storage: string interner should not panic in release

### PR DESCRIPTION
**Pull request by cstyan**
Need to dig a bit more to find the call stack that lead to the panic but this will mean we don't panic at least.

@brian-brazil @beorn-

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cc @tomwilkie

### CODE REVIEW

File: storage/remote/intern.go [hunk 1]
```diff
@@ -73,7 +83,7 @@ func (p *pool) release(s string) {
 	p.mtx.RUnlock()
 
 	if !ok {
-		panic("released unknown string")
+		p.noReferenceReleases.Inc()
```
Comment (brian-brazil) [inline storage/remote/intern.go]: Should we return here?

File: storage/remote/intern.go [hunk 2]
```diff
@@ -38,6 +42,12 @@ type entry struct {
 func newPool() *pool {
 	return &pool{
 		pool: map[string]*entry{},
+		noReferenceReleases: prometheus.NewCounter(prometheus.CounterOpts{
```
Comment (simonpasquier) [inline storage/remote/intern.go]: The metric isn't registered anywhere?

**Comment by cstyan** [comment]
PTAL

cc @simonpasquier I needed the `testutil` package from `client_golang`, which wasn't already vendored. I ran `go get github.com/prometheus/client_golang/prometheus/testutil@65d3a96fbaa7c8c9535d7133d6d98cd50eed4db8` and then `go mod tidy && go mod vendor`. Let me know if it doesn't look correct. I've noticed that our `go.mod` says that we're using `client_golang` at `v0.9.3` for some commit even though the latest actual release was `v0.9.2`. I assume go mod updates the version in the files itself.

**Comment by cstyan** [comment]
fixes #5480

**Comment by simonpasquier** [comment]
> I needed the testutil package from client_golang, which wasn't already vendored. I ran go get github.com/prometheus/client_golang/prometheus/testutil@65d3a96fbaa7c8c9535d7133d6d98cd50eed4db8 and then go mod tidy && go mod vendor. 

You wouldn't even need to `go get .../client_golang` as it was already present. Importing the testutil package and running `go mod tidy/go mod vendor` would have been enough to update the vendor directory.

> I've noticed that our go.mod says that we're using client_golang at v0.9.3 for some commit even though the latest actual release was v0.9.2. I assume go mod updates the version in the files itself.

This is because we're referencing a commit after the last known tag, see https://tip.golang.org/cmd/go/#hdr-Pseudo_versions.

File: storage/remote/intern.go [hunk 3]
```diff
@@ -36,9 +40,17 @@ type entry struct {
...
+	p := &pool{
 		pool: map[string]*entry{},
+		noReferenceReleases: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: namespace,
+			Subsystem: subsystem,
+			Name:      "string_interner_zero_reference_releases_total",
+			Help:      "The number of times release has been called for strings that are not interned.",
+		}),
 	}
+	prometheus.Register(p.noReferenceReleases)
```
Comment (brian-brazil) [inline storage/remote/intern.go]: This is unsafe, both the metric and the registration should be at file level.

**Comment by cstyan** [comment]
I'd like to make the interner and it's metric not package level variables but I'll do that later in another PR

**Comment by brian-brazil** [comment]
Thanks!

**Comment by beorn-** [comment]
Thank you very much !

**Pull request merged**